File: swf-emitter.ts/src/lib/emitters/gradient.ts
import { Incident } from "incident";
import { Uint2, Uint8 } from "semantic-types";
import { ColorSpace, ColorStop, Gradient, GradientSpread, MorphColorStop, MorphGradient } from "swf-tree";
import { ByteStream } from "../stream";
import { emitSRgb8, emitStraightSRgba8 } from "./basic-data-types";

const GRADIENT_SPREAD_TO_CODE: Map<GradientSpread, Uint2> = new Map([
  [GradientSpread.Pad, 0 as Uint2],
  [GradientSpread.Reflect, 1 as Uint2],
  [GradientSpread.Repeat, 2 as Uint2],
]);

const COLOR_SPACE_TO_CODE: Map<ColorSpace, Uint2> = new Map([
  [ColorSpace.LinearRgb, 1 as Uint2],
  [ColorSpace.SRgb, 0 as Uint2],
]);

export function emitColorStop(byteStream: ByteStream, value: ColorStop, withAlpha: boolean): void {
  byteStream.writeUint8(value.ratio);
  if (withAlpha) {
    emitStraightSRgba8(byteStream, value.color);
  } else {
    emitSRgb8(byteStream, value.color);
  }
}

export function emitGradient(byteStream: ByteStream, value: Gradient, withAlpha: boolean): void {
  const spreadCode: Uint2 | undefined = GRADIENT_SPREAD_TO_CODE.get(value.spread);
  const colorSpaceCode: Uint2 | undefined = COLOR_SPACE_TO_CODE.get(value.colorSpace);
  if (spreadCode === undefined) {
    throw new Incident("UnexpectedSpread");
  }
  if (colorSpaceCode === undefined) {
    throw new Incident("UnexpectedColorSpace");
  }

  const flags: Uint8 = 0
    | ((value.colors.length & 0x0f) << 0)
    | ((colorSpaceCode & 0x03) << 4)
    | ((spreadCode & 0x03) << 6);
  byteStream.writeUint8(flags);

  for (const colorStop of value.colors) {
    emitColorStop(byteStream, colorStop, withAlpha);
  }
}

export function emitMorphColorStop(byteStream: ByteStream, value: MorphColorStop, withAlpha: boolean): void {
  emitColorStop(byteStream, {ratio: value.ratio, color: value.color}, withAlpha);
  emitColorStop(byteStream, {ratio: value.morphRatio, color: value.morphColor}, withAlpha);
}

export function emitMorphGradient(byteStream: ByteStream, value: MorphGradient, withAlpha: boolean): void {
  const spreadCode: Uint2 | undefined = GRADIENT_SPREAD_TO_CODE.get(value.spread);
  const colorSpaceCode: Uint2 | undefined = COLOR_SPACE_TO_CODE.get(value.colorSpace);
  if (spreadCode === undefined) {
    throw new Incident("UnexpectedSpread");
  }
  if (colorSpaceCode === undefined) {
    throw new Incident("UnexpectedColorSpace");
  }

  const flags: Uint8 = 0
    | ((value.colors.length & 0x0f) << 0)
    | ((colorSpaceCode & 0x03) << 4)
    | ((spreadCode & 0x03) << 6);
  byteStream.writeUint8(flags);

  for (const colorStop of value.colors) {
    emitMorphColorStop(byteStream, colorStop, withAlpha);
  }
}

File: rs/src/gradient.rs
use std::convert::TryFrom;
use std::io;

use swf_types as ast;

use crate::basic_data_types::{emit_s_rgb8, emit_straight_s_rgba8};
use crate::primitives::emit_u8;

fn gradient_spread_to_code(value: ast::GradientSpread) -> u8 {
  match value {
    ast::GradientSpread::Pad => 0,
    ast::GradientSpread::Reflect => 1,
    ast::GradientSpread::Repeat => 2,
  }
}

fn color_space_to_code(value: ast::ColorSpace) -> u8 {
  match value {
    ast::ColorSpace::LinearRgb => 1,
    ast::ColorSpace::SRgb => 0,
  }
}

pub(crate) fn emit_gradient<W: io::Write + ?Sized>(
  writer: &mut W,
  value: &ast::Gradient,
  with_alpha: bool,
) -> io::Result<()> {
  assert!(value.colors.len() <= 0x0f);
  let flags: u8 = 0
    | ((u8::try_from(value.colors.len()).unwrap() & 0x0f) << 0)
    | ((gradient_spread_to_code(value.spread) & 0b11) << 4)
    | ((color_space_to_code(value.color_space) & 0b11) << 6);
  emit_u8(writer, flags)?;

  for color_stop in &value.colors {
    emit_color_stop(writer, color_stop, with_alpha)?;
  }

  Ok(())
}

pub(crate) fn emit_color_stop<W: io::Write + ?Sized>(
  writer: &mut W,
  value: &ast::ColorStop,
  with_alpha: bool,
) -> io::Result<()> {
  emit_u8(writer, value.ratio)?;
  if with_alpha {
    emit_straight_s_rgba8(writer, value.color)
  } else {
    assert!(value.color.a == u8::max_value());
    emit_s_rgb8(
      writer,
      ast::SRgb8 {
        r: value.color.r,
        g: value.color.g,
        b: value.color.b,
      },
    )
  }
}

pub(crate) fn emit_morph_gradient<W: io::Write + ?Sized>(writer: &mut W, value: &ast::MorphGradient) -> io::Result<()> {
  assert!(value.colors.len() <= 0x0f);
  let flags: u8 = 0
    | ((u8::try_from(value.colors.len()).unwrap() & 0x0f) << 0)
    | ((gradient_spread_to_code(value.spread) & 0b11) << 4)
    | ((color_space_to_code(value.color_space) & 0b11) << 6);
  emit_u8(writer, flags)?;

  for color_stop in &value.colors {
    emit_morph_color_stop(writer, color_stop)?;
  }

  Ok(())
}

pub(crate) fn emit_morph_color_stop<W: io::Write + ?Sized>(
  writer: &mut W,
  value: &ast::MorphColorStop,
) -> io::Result<()> {
  emit_color_stop(
    writer,
    &ast::ColorStop {
      ratio: value.ratio,
      color: value.color,
    },
    true,
  )?;
  emit_color_stop(
    writer,
    &ast::ColorStop {
      ratio: value.morph_ratio,
      color: value.morph_color,
    },
    true,
  )
}
